solana/restaking/programs/restaking/src/lib.rs

```diff
@@ -0,0 +1,488 @@
+    pub receipt_token_mint: Box<Account<'info, Mint>>,
+    #[account(mut, token::mint = receipt_token_mint, token::authority = withdrawer)]
+    pub receipt_token_account: Box<Account<'info, TokenAccount>>,
+
+    pub guest_chain_program: Program<'info, SolanaIbc>,
+    pub token_program: Program<'info, Token>,
+    pub system_program: Program<'info, System>,
+    pub metadata_program: Program<'info, Metadata>,
+    pub rent: Sysvar<'info, Rent>,
+    #[account(
+        mut,
+        seeds = [
+            b"metadata".as_ref(),
+            metadata_program.key().as_ref(),
+            receipt_token_mint.key().as_ref(),
+            b"edition".as_ref(),
+        ],
+        bump,
+        seeds::program = metadata_program.key()
+    )]
+    /// CHECK:
+    pub master_edition_account: UncheckedAccount<'info>,
+    #[account(
+        mut,
+        seeds = [
+            b"metadata".as_ref(),
+            metadata_program.key().as_ref(),
+            receipt_token_mint.key().as_ref(),
+        ],
+        bump,
+        seeds::program = metadata_program.key()
+    )]
+    /// CHECK:
+    pub nft_metadata: UncheckedAccount<'info>,
+}
+
+#[derive(Accounts)]
+pub struct UpdateTokenWhitelist<'info> {
+    #[account(mut)]
+    pub admin: Signer<'info>,
+
+    #[account(mut, seeds = [STAKING_PARAMS_SEED], bump)]
+    pub staking_params: Account<'info, StakingParams>,
+}
+
+#[derive(Accounts)]
+pub struct Claim<'info> {
+    #[account(mut)]
+    pub claimer: Signer<'info>,
+
+    #[account(mut, seeds = [VAULT_PARAMS_SEED, receipt_token_mint.key().as_ref()], bump)]
+    pub vault_params: Box<Account<'info, Vault>>,
+    #[account(mut, seeds = [STAKING_PARAMS_SEED, TEST_SEED], bump, has_one = rewards_token_mint)]
+    pub staking_params: Box<Account<'info, StakingParams>>,
+
+    #[account(mut, seeds = [CHAIN_SEED], bump, seeds::program = guest_chain_program.key())]
+    pub guest_chain: Box<Account<'info, ChainData>>,
+
+    pub rewards_token_mint: Box<Account<'info, Mint>>,
+    #[account(mut, token::mint = rewards_token_mint, token::authority = claimer)]
+    pub depositor_rewards_token_account: Box<Account<'info, TokenAccount>>,
+
+    #[account(mut, seeds = [REWARDS_SEED, TEST_SEED], bump, token::mint = rewards_token_mint, token::authority = staking_params)]
+    pub platform_rewards_token_account: Box<Account<'info, TokenAccount>>,
+
+    #[account(mut, mint::decimals = 0)]
+    pub receipt_token_mint: Box<Account<'info, Mint>>,
+    #[account(mut, token::mint = receipt_token_mint, token::authority = claimer)]
+    pub receipt_token_account: Box<Account<'info, TokenAccount>>,
+
+    pub guest_chain_program: Program<'info, SolanaIbc>,
+    pub token_program: Program<'info, Token>,
+    pub system_program: Program<'info, System>,
+    pub rent: Sysvar<'info, Rent>,
+}
+
+#[account]
+#[derive(InitSpace)]
+pub struct StakingParams {
+    pub admin: Pubkey,
+    #[max_len(20)]
+    pub whitelisted_tokens: Vec<Pubkey>,
+    pub bounding_timestamp: u64,
```

Comment:
Let’s do i64 so we don’t have to convert between type used by Solana.  Also, let’s suffix it with unit.

```suggestion
    pub bounding_timestamp_sec: i64,
```